nixos/modules/services/backup/rustic.nix

```diff
@@ -0,0 +1,445 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  tomlFormat = pkgs.formats.toml { };
+
+  cfg = config.services.rustic;
+  configFiles = lib.mapAttrs (k: v: tomlFormat.generate "rustic-${k}.toml" v) cfg.profiles;
+
+  commonOpts =
+    v:
+    if v ? files then
+      v.files
+    else if v ? command then
+      v.command
+    else if v ? postgres then
+      v.postgres
+    else
+      throw "unexpected: backup has none of the authorized types";
+
+  backupType =
+    v:
+    if v ? files then
+      "files"
+    else if v ? command then
+      "command"
+    else if v ? postgres then
+      "postgres"
+    else
+      throw "unexpected: backup has none of the authorized types";
+
+  backupConfigToCmd =
+    k: v:
+    let
+      o = commonOpts v;
+      profileArgs = lib.concatMapStrings (s: " -P \"${s}\"") o.useProfiles;
+      extraArgs = lib.concatMapStrings (s: " \"${s}\"") o.extraArgs;
+      genericArgs = profileArgs + extraArgs;
+
+      sourcesArgs = lib.concatMapStrings (s: " ${s}") v.files.sources;
+      asPathArg = lib.optionalString (v.files.asPath != null) " --as-path \"${v.files.asPath}\"";
+      ifFilesArgs = lib.optionalString (v ? files) (sourcesArgs + asPathArg);
+
+      commandArg = " --stdin-command \"${v.command.command}\" -"; # stdin-command requires adding `-` as a source
+      stdinFilenameArg = lib.optionalString (
+        v.command.filename != null
+      ) " --stdin-filename \"${v.command.filename}\"";
+      ifCommandArgs = lib.optionalString (v ? command) (commandArg + stdinFilenameArg);
+
+      pipeJsonIntoCmd =
+        if o.pipeJsonInto ? nothing then
+          ""
+        else if o.pipeJsonInto ? command then
+          " --json | ${o.pipeJsonInto.command} \"$1\""
+        else if o.pipeJsonInto ? prometheus then
+          " --json | ${prometheusWriteScript o.pipeJsonInto.prometheus.nodeExporterCollectorFolder} \"$1\""
+        else
+          throw "unexpected: pipeJsonInto has none of the authorized types";
+    in
+    if v ? postgres then
+      let
+        systemctl = "${config.systemd.package}/bin/systemctl";
+      in
+      pkgs.writeScript "rustic-postgres-starter-${k}" ''
+        #!${pkgs.bash}/bin/bash
+        set -euo pipefail
+
+        ${systemctl} start --no-block rustic-postgres-globals-${k}.service
+
+        ${pkgs.sudo}/bin/sudo -u postgres \
+          ${config.services.postgresql.package}/bin/psql \
+          -c 'SELECT datname FROM pg_database' \
+          --csv \
+          | tail -n +2 \
+          | grep -Ev '^template0$' \
+          | xargs -I {} \
+          ${systemctl} start --no-block rustic-postgres-db-${k}@{}.service
+      ''
+    else
+      pkgs.writeScript "rustic-backup-${k}" ''
+        #!${pkgs.bash}/bin/bash
+        set -euo pipefail
+
+        TZ="" ${cfg.package}/bin/rustic backup${ifFilesArgs + ifCommandArgs + genericArgs + pipeJsonIntoCmd}
```

Comment:
Why clear TZ?

Reply:
This is required for the JSON output to be relatively easily parsable: otherwise, the dates in the JSON output are given in a TZ-local timezone, and makes parsing unbelievably hard.

Thinking more about it, maybe it'd make sense to add that only when `pipeJsonInto` is non-empty?